flanksource/karina
# Use restic to backup/restore PostgreSQL Cluster

### Description

Done:
- Scheduled DB backup using restic (via CronJob) is implemented and tested
- On-demand DB Backup using restic (via `karina db backup`) is implemented and tested
- Add a new command/flags to list all backups for a specific database (and can optionally specify a bucket): `karina db backup list [--bucket BucketName] [-q] [-n NumberOfBackupsToReturn]`
- Move most of the shared configurations for Postgres Clusters to `secret/postgres-operator/postgres-operator-cluster-environment`. This will become a single source of truth for all PG Clusters
- Implemented backup retention
- Implement `karina db restore` to restore from restic backups
- Update postgres operator docs
- Update backup image after https://github.com/flanksource/postgres-backups/pull/1 is merged

### Dependencies

- https://github.com/flanksource/postgres-backups/pull/1
- https://github.com/flanksource/kommons/pull/32
- https://github.com/flanksource/template-operator-library/pull/10
- https://github.com/flanksource/kommons/pull/33
- https://github.com/flanksource/template-operator-library/pull/11

### Breaking Change


### Testing Notes

**What I did:**
- My `karina.yml`:
```yaml
domain: 127.0.0.1.nip.io
name: test-cluster
kind:
  workerCount: 3
localPath:
  disabled: false
calico:
  version: v3.8.2
kubernetes:
  version: v1.18.6
  masterIP: localhost
  containerRuntime: containerd
ca:
  cert: .certs/root-ca.crt
  privateKey: .certs/root-ca.key
  password: foobar
ingressCA:
  cert: .certs/ingress-ca.crt
  privateKey: .certs/ingress-ca.key
  password: foobar
mongodbOperator:
  version: v1.6.0
keptn:
  version: v0.7.3
templateOperator:
  version: v0.1.11
  syncPeriod: 5s
postgresOperator:
  version: v1.3.4.flanksource.1
  backupPassword: password123456
  defaultBackupBucket: long-test-pg-backup
  defaultBackupRetention:
    keepLast: 5
    keepHourly: 2
    keepDaily: 1
minio:
  version: RELEASE.2020-09-02T18-19-50Z
  access_key: minio
  secret_key: minio123
  # 1 or non HA, otherwise must be in multiples of 4
  replicas: 1
s3:
  access_key: minio
  secret_key: minio123
  endpoint: http://minio.minio.svc:9000
  skipTLSVerify: true
  usePathStyle: true
canaryChecker:
  version: v0.15.1
```
- Provision local env: `karina provision kind-cluster`
- Deploy necessary runtimes: `karina deploy phases --crds --calico --base --csi --template-operator --postgres-operator --minio --canary`
- My `db.yaml`
```yaml
apiVersion: db.flanksource.com/v1
kind: PostgresqlDB
metadata:
  name: test-db
  namespace: postgres-operator
spec:
  backup:
#    bucket: test-pg-backup
    schedule: "*/5 * * * *"
    retention:
      keepDaily: 30
  cpu: 200m
  memory: 1Gi
  replicas: 1
  storage:
    size: 1Gi
  parameters:
    archive_mode: "on"
    archive_timeout: 60s
    log_destination: stderr
    max_connections: "600"
    shared_buffers: 2048MB
```
- Deploy the db: `k apply -f db.yml`
- Wait for the backup to be created via CronJob
- Run a backup on demand with `karina db backup --name postgres-test-db`, the job should run successfully
- List the list of backups for the db with `karina db backup list --name postgres-test-db`
```
BACKUP PATH                                                                                   TIME                          AGE
restic:s3:http://minio.minio.svc:9000/test-pg-backup/postgres-test-db/2021-04-02.070007.sql   2021-04-04 07:00:08 +00 UTC   2 minutes
restic:s3:http://minio.minio.svc:9000/test-pg-backup/postgres-test-db/2021-04-02.065506.sql   2021-04-04 06:55:07 +00 UTC   7 minutes
restic:s3:http://minio.minio.svc:9000/test-pg-backup/postgres-test-db/2021-04-02.065343.sql   2021-04-04 06:53:43 +00 UTC   8 minutes
...
```

**Test restoring logical backup into a different PG cluster**
- My `db_restore.yml`
```yaml
apiVersion: db.flanksource.com/v1
kind: PostgresqlDB
metadata:
  name: test-db-restore
  namespace: postgres-operator
spec:
  backup:
    bucket: test-pg-restore
    schedule: "*/5 * * * *"
    retention:
      keepDaily: 30
  cpu: 200m
  memory: 1Gi
  replicas: 1
  storage:
    size: 1Gi
  parameters:
    archive_mode: "on"
    archive_timeout: 60s
    log_destination: stderr
    max_connections: "600"
    shared_buffers: 2048MB
```
- Deploy the new pg cluster with `k apply -f db_restore.yml`
- Shell to the first DB and create some data:
```
create table long_test(id serial primary key);
insert into long_test values (1);
insert into long_test values (2);
insert into long_test values (3);
```
- Create an on-demand logical backup with `karina db backup --name postgres-test-db`
- List the backup with `karina db backup list --name postgres-test-db`, copy the backup path of the latest backup
- Run this to restore this backup to the second cluster `karina db restore BACKUP_PATH_COPIED_FROM_LAST_STEP --name postgres-test-db-restore`
- Shell to the second pg cluster and test if the data has been restored:
```
select * from long_test;
```

**How you can replicate my testing:**
Same

### **Links**

Implement half of #589 

Closes #462
